src/kpro.erl

```diff
@@ -138,764 +213,302 @@
+  kpro_brokers:discover_partition_leader(Connection, Topic, Partition, Timeout).
+
+%% @see connect_controller/3.
+-spec connect_controller(connection() | [endpoint()], conn_config()) ->
+        {ok, connection()} | {error, any()}.
+connect_controller(Bootstrap, ConnConfig) ->
+  connect_controller(Bootstrap, ConnConfig, #{}).
+
+%% @doc Conect to the controller broker of the kafka cluster.
+-spec connect_controller(connection() | [endpoint()], conn_config(),
+                         #{timeout => timeout()}) ->
+        {ok, connection()} | {error, any()}.
+connect_controller(Bootstrap, ConnConfig, Opts) ->
+  kpro_brokers:connect_controller(Bootstrap, ConnConfig, Opts).
+
+%% @doc Connect to group or transaction coordinator.
+%% If the first arg is not a connection pid but a list of bootstraping
+%% endpoints, it will frist try to connect to any of the nodes
+%% NOTE: 'txn' type only applicable to kafka 0.11 or later
+-spec connect_coordinator(connection() | [endpoint()], conn_config(),
+                          #{ type => coordinator_type()
+                           , id => binary()
+                           , timeout => timeout()
+                           }) -> {ok, connection()} | {error, any()}.
+connect_coordinator(Bootstrap, ConnConfig, Args) ->
+  kpro_brokers:connect_coordinator(Bootstrap, ConnConfig, Args).
+
+%% @doc Discover group or transactional coordinator.
+%% An implicit step performed in `connect_coordinator'.
+%% This is useful when the caller wants to re-use already established
+%% towards the discovered endpoint.
+-spec discover_coordinator(connection(), coordinator_type(),
+                           group_id() | transactional_id(), timeout()) ->
+        {ok, endpoint()} | {error, any()}.
+discover_coordinator(Connection, Type, Id, Timeout) ->
+  kpro_brokers:discover_coordinator(Connection, Type, Id, Timeout).
+
+%% @doc Qury API versions using the given `kpro_connection' pid.
+-spec get_api_versions(connection()) ->
+        {ok, vsn_ranges()} | {error, any()}.
+get_api_versions(Connection) ->
+  kpro_brokers:get_api_versions(Connection).
+
+%% @doc Get version range for the given API.
+-spec get_api_vsn_range(connection(), api()) ->
+        {ok, vsn_range()} | {error, any()}.
+get_api_vsn_range(Connection, API) ->
+  kpro_brokers:get_api_vsn_range(Connection, API).
+
+%% @doc Find field value in a struct, raise an 'error' exception if not found.
+-spec find(field_name(), struct()) -> field_value().
 find(Field, Struct) ->
-  case lists:keyfind(Field, 1, Struct) of
-    {_, Value} -> Value;
-    false -> erlang:throw({no_such_field, Field})
-  end.
+  kpro_lib:find(Field, Struct, {no_such_field, Field}).
 
 %% @doc Find field value in a struct, reutrn default if not found.
 -spec find(field_name(), struct(), field_value()) -> field_value().
 find(Field, Struct, Default) ->
   try
     find(Field, Struct)
   catch
-    throw : {no_such_field, _} ->
+    error : {no_such_field, _} ->
```

Comment:
It's not nice to have an API that forces an exception, and then in some cases catch it and turn it into a OOB value.